torch/csrc/jit/codegen/cuda/kernel_cache.h

```diff
@@ -21,16 +23,63 @@ namespace cuda {
 // TODO: Figure out how we want to cache based on heuristics, should probably
 // use something similar. Heuristics may also return a LaunchParams object.
 // TODO: Validate it is included in the string.
+
 class FusionExecutorCache {
  public:
-  FusionExecutor* getExecutor() const {
-    return entry;
-  }
+  // FusionExecutorCache(Fusion* fusion, CompileOptions options);
+  FusionExecutorCache(std::unique_ptr<Fusion>&& fusion, at::Device device);
+
+  std::vector<at::Tensor> runFusionWithInputs(
+      const at::ArrayRef<IValue>& inputs);
+
+ private:
+  at::Device device_;
+  std::unique_ptr<Fusion> fusion_;
+  std::vector<std::unique_ptr<FusionExecutor>> fusion_executor_cache_;
+};
+
+class GraphCache {
+ public:
+  GraphCache(std::shared_ptr<Graph> graph);
 
-  FusionExecutorCache(Fusion* fusion, CompileOptions options);
+  std::vector<at::Tensor> runGraphWithInputs(
+      const at::ArrayRef<IValue>& inputs);
 
  private:
-  FusionExecutor* entry = nullptr;
+  // TODO: place holder with naive implementation for now.
+  struct InputsRequirement {
+    c10::optional<at::Device> device_;
+    at::DimVector input_permutation_;
+    at::DimVector output_permutation_;
+    // TODO: TensorTypePtr is not very easy to work with.
+    // c10::nullopt to take place of non-tensor type;
+    std::vector<c10::optional<at::TensorTypePtr>> vec_optional_ttp;
+
+    InputsRequirement(
+        const std::shared_ptr<Graph>& graph,
+        const std::vector<size_t>& reduction_axes);
+    InputsRequirement(
+        const at::ArrayRef<IValue>& inputs,
+        const std::vector<size_t>& reduction_axes);
+
+    // bool operator==(const InputsRequirement& other);
+    bool complyWith(const InputsRequirement& expect);
+
+    bool requiresPermutation();
+  };
+
+  FusionExecutorCache* createFusionExecutorCache(
+      const InputsRequirement& input_stack);
+
+  // Computation graph;
```

Comment:
nit: please add another `private:` to separate private methods from private state